EventGenerator/src/CaloCalibGun_module.cc

```diff
@@ -101,138 +118,120 @@ namespace mu2e {
     produces<mu2e::PrimaryParticle>();
-    produces <MCTrajectoryCollection>();
 
-    if (_mean < 0) throw cet::exception("RANGE") << "CaloCalibGun.mean must be non-negative "<< '\n';
   }
 
+  void CaloCalibGun::beginRun(art::Run&){
+      const DiskCalorimeter *_cal = GeomHandle<DiskCalorimeter>().get();
+
+      _pipeRadius      = _cal->caloInfo().getDouble("pipeRadius");
+      _pipeTorRadius   = _cal->caloInfo().getVDouble("pipeTorRadius");
+      _zPipeCenter     = _cal->disk(_nDisk).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(_nDisk).geomInfo().size().z()/2.0-_pipeRadius);
+      _nPipes = _cal->caloInfo().getInt("nPipes");
+   
+      //Define the parameters of the pipes:
+      phi_lbd = _cal->caloInfo().getVDouble("largeTorPhi");
+      phi_sbd = _cal->caloInfo().getVDouble("smallTorPhi");
+      phi_end = _cal->caloInfo().getVDouble("straightEndPhi");
+      ysmall = _cal->caloInfo().getVDouble("yposition");
+      radSmTor = _cal->caloInfo().getDouble("radSmTor");
+      xsmall = _cal->caloInfo().getDouble("radSmTor");
+      xdistance = _cal->caloInfo().getDouble("xdistance");
+      rInnerManifold = _cal->caloInfo().getDouble("rInnerManifold");
+      
+
+  }
   //================================================================
+  
   void CaloCalibGun::produce(art::Event& event) {
     std::unique_ptr<GenParticleCollection> output(new GenParticleCollection);
     PrimaryParticle primaryParticles;
-    MCTrajectoryCollection mctc;
-
-    const DiskCalorimeter *_cal = GeomHandle<DiskCalorimeter>().get();
-    //_nPipes          = _cal->caloInfo().getInt("nPipes");
-    _pipeRadius      = _cal->caloInfo().getDouble("pipeRadius");
-    _pipeTorRadius   = _cal->caloInfo().getVDouble("pipeTorRadius");
-    _randomRad       = _cal->caloInfo().getVDouble("pipeTorRadius");
-    _zPipeCenter     = _cal->disk(_nDisk).geomInfo().origin()-CLHEP::Hep3Vector(0,0,_cal->disk(_nDisk).geomInfo().size().z()/2.0-_pipeRadius);//disk =1
-    // we normalize to the volume of the pipe (proportional to 2*pi*R if they have all the same radius) to draw a random number from which to generate the photons TODO - is this used?
-    double sumR(0);
-    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {sumR+=d; d = sumR; });
-    std::for_each(_randomRad.begin(), _randomRad.end(), [&](double& d) {d /= sumR;});
- 
-    //Define the parameters of the pipes:
+
+    double xpipe, ypipe, zpipe;
+    //Pick position - find either 0,1 - these are indices of the sign list (so 0=-1, 1=+1)
+    int xsn = round(_randFlat.fire()); 
+    int ysn = round(_randFlat.fire());
+
+    // pick a random theta, between 0 and 2*pi:
+    double theta = _randFlat.fire() * 2.0 * CLHEP::pi;
+    // pick a random point on the radius
+    double pipeR = _pipeRadius * sqrt(_randFlat.fire());
+    // find the z position based on above:
+    zpipe = pipeR*sin(theta);
+
+    // select an index from list of pipes:
+    unsigned int idx = int(5*_randFlat.fire())+1;
```

Comment:
Don't you want an index in the range 0 to 4, not 1 to 5?  You are indexing into std::vector with dimension 5.

Reply:
Thanks for fixing this.